server/devices.go

```diff
@@ -10,158 +10,146 @@ import (
 			statusCode = http.StatusAccepted
-			msg = ServerMsg{Message: "Device not found"}
+			msg, _ = json.Marshal(ServerMsg{Message: "Device not found"})
 			return
 		}
 		statusCode = http.StatusOK
-		json.NewEncoder(w).Encode(ret)
+		msg, _ = json.Marshal(ret)
 	case "POST":
 		// Add a new device.
 		newDevice, err := clientdevices.JsonReq2Device(req)
-
 		if err != nil {
 			statusCode = http.StatusBadRequest
-			msg = ServerMsg{Message: fmt.Sprintf("Error decoding json: %s", err)}
+			msg, _ = json.Marshal(ServerMsg{Message: fmt.Sprintf("Error decoding json: %s", err)})
 			return
 		}
+
+		if newDevice.Name == "" {
+			statusCode = http.StatusBadRequest
+			msg, _ = json.Marshal(ServerMsg{Message: "Body missing required fields"})
+			return
+		}
+
 		err = s.DeivcesClient.AddDevice(newDevice)
 		if err != nil {
+			if err.Error() == "device already exists" {
+				statusCode = http.StatusBadRequest
+				msg, _ = json.Marshal(ServerMsg{Message: "device already exists"})
+				return
+			}
 			statusCode = http.StatusInternalServerError
 			log.WithFields(log.Fields{"new_device": newDevice}).Error("Error adding new device")
 			return
 		}
 		statusCode = http.StatusOK
-		msg = ServerMsg{Message: "Adding new device successful"}
+		msg, _ = json.Marshal(ServerMsg{Message: "Added new device"})
 	case "PUT":
 		// Update an existing record.
 		newDevice, err := clientdevices.JsonReq2Device(req)
 		if err != nil {
 			statusCode = http.StatusBadRequest
-			msg = ServerMsg{Message: fmt.Sprintf("Error decoding json: %s", err)}
+			msg, _ = json.Marshal(ServerMsg{Message: fmt.Sprintf("Error decoding json: %s", err)})
 			return
 		}
 
 		if newDevice.Name == "" {
-			msg = ServerMsg{Message: "You must specify name when trying to update device"}
+			msg, _ = json.Marshal(ServerMsg{Message: "You must specify name when trying to update device"})
 			statusCode = http.StatusBadRequest
 			return
 		}
+
 		err = s.DeivcesClient.UpdateDevice(newDevice)
 		if err != nil {
+			if err.Error() == "device not found" {
+				statusCode = http.StatusNotFound
+				msg, _ = json.Marshal(ServerMsg{Message: "Device not found"})
+				return
+			}
 			statusCode = http.StatusInternalServerError
 			log.WithFields(log.Fields{"new_device": newDevice}).Error("Error updating device")
 			return
 		}
+
 		statusCode = http.StatusOK
-		msg = ServerMsg{Message: "Update successful"}
+		msg, _ = json.Marshal(ServerMsg{Message: "Update successful"})
 	case "DELETE":
 		// Remove the record.
-		urlPar, _ := url.Parse(req.RequestURI)
-		qmap, _ := url.ParseQuery(urlPar.RawQuery)
-		err := s.DeivcesClient.RemoveDeviceByName(qmap["device"][0])
+		qmap, _ := url.ParseQuery(req.URL.RawQuery)
+		deviceName := qmap.Get("device")
+
+		err := s.DeivcesClient.RemoveDeviceByName(deviceName)
 
 		if err != nil {
-			statusCode = http.StatusBadRequest
-			msg = ServerMsg{Message: fmt.Sprintf("Couldn't Delete device: %s", qmap["device"][0])}
+			if err.Error() == "device not found" {
```

Comment:
@rsikdar  thoughts on this?

Reply:
DeivcesClient could have a constant or function to check for it
```
if devicesClient.IsErrNotFound(err):
```

otherwise, I would recommend doing string.Contains at the very least instead of exact comparison.